chat-history: Mark visible messages as viewed

marhkb opened the pull request.

marhkb committed:
chat: Add property `last-read-inbox-message-id`

melix99 commented on src/session/content/chat_history.rs:
```diff
@@ -223,4 +228,148 @@ impl ChatHistory {
+
+                imp.view_messages_source_id.take();
+
+                let list =  imp.list_view.observe_children();
```
I don't think this is the right approach. This function is [heavily discouraged by gtk docs](https://docs.gtk.org/gtk4/method.Widget.observe_children.html) because, other than returning the list of children, it also tracks them using the listmodel signals, which slowdowns things. I think we should use first_child() and next_sibling() functions to iterate children, like we do [here](https://github.com/melix99/telegrand/blob/6c3372207d16bb0a3d5853751cfca12bedd92b9c/src/session/content/message_row/base.rs#L71).

melix99 replied:
Btw, I've checked and observe_children() also iterates using next_sibling(), but it does that using a custom listmodel + tracks children, so it does extra overhead which we don't need anyway.

marhkb replied:
This is indeed bad. And binary search is also counterproductive as this is a linked list without random access. So I think just iterating the children sequentially is the best thing we can do.

marhkb committed:
chat: Update property `last-read-inbox-message-id`

marhkb committed:
Chat-history: Mark visible messages as viewed

This includes both the bottommost regular message and sponsored
messages.

A regular message is considered as being viewed if parts of it are
either visible inside the specified list view or if the user needs to
scroll up in order to see it.

Inside a list view `L` a sponsored message `S` is considered as being
viewed if `L ∩ S = S` applies.
